src/main/java/uk/gov/pay/ledger/util/csv/CSVMessageBodyWriter.java

```diff
@@ -36,11 +38,16 @@ public void writeTo(List data, Class aClass, Type type, Annotation[] annotations
         if (data != null && !data.isEmpty()) {
             CsvMapper mapper = new CsvMapper();
 
-            // rank properties by class property order instead of alphabetically
             mapper.disable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY);
-            Object sample = data.get(0);
-            CsvSchema schema = mapper.schemaFor(sample.getClass()).withHeader();
+            mapper.configure(JsonGenerator.Feature.IGNORE_UNKNOWN, true);
+
+            CsvSchema.Builder builder = CsvSchema.builder();
+            Map<String, Object> headerRow = (Map<String, Object>) data.get(data.size() - 1); // get last record which is header
```

Comment:
Feels a bit hacky.
It would be cleaner to have an object having headers separate from data (two different properties) rather than having this assumption that the last record is a header (someone can change it by accident and the whole report looks interesting).

Reply:
Agree, but can't have headers separately as we can only pass `data` list to CSVMessageBodyWriter and as we need to generate headers from data, headers are tied to data set.

Other way is to explore other libraries for CSV data generation, but as discussed we would like to see the performance when using Maps. With the POJO we could get up 35K records successfully and failed to get 70K odd.